File: frescophotoview/src/main/java/com/zml/frescophotoview/GestureListenerWrapper.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.zml.frescophotoview;

import android.view.GestureDetector;
import android.view.MotionEvent;


/** Wrapper for SimpleOnGestureListener as GestureDetector does not allow changing its listener. */
public class GestureListenerWrapper extends GestureDetector.SimpleOnGestureListener {

  private GestureDetector.SimpleOnGestureListener mOuterListener;
  private DefaultGestureHandler mDefaultGestureHandler;

  public GestureListenerWrapper(DefaultGestureHandler defaultGestureHandler) {
    mDefaultGestureHandler = defaultGestureHandler;
  }

  public void setListener(GestureDetector.SimpleOnGestureListener listener) {
    mOuterListener = listener;
  }

  @Override
  public void onLongPress(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onLongPress(e);
    }
    mDefaultGestureHandler.onLongPress(e);
  }

  @Override
  public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
    if (mOuterListener != null) {
      mOuterListener.onScroll(e1, e2, distanceX, distanceY);
    }
    return mDefaultGestureHandler.onScroll(e1, e2, distanceX, distanceY);
  }

  @Override
  public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
    if (mOuterListener != null) {
      mOuterListener.onFling(e1, e2, velocityX, velocityY);
    }
    return mDefaultGestureHandler.onFling(e1, e2, velocityX, velocityY);
  }

  @Override
  public void onShowPress(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onShowPress(e);
    }
    mDefaultGestureHandler.onShowPress(e);
  }

  @Override
  public boolean onDown(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onDown(e);
    }
    return mDefaultGestureHandler.onDown(e);
  }

  @Override
  public boolean onDoubleTap(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onDoubleTap(e);
    }
    return mDefaultGestureHandler.onDoubleTap(e);
  }

  @Override
  public boolean onDoubleTapEvent(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onDoubleTapEvent(e);
    }
    return mDefaultGestureHandler.onDoubleTapEvent(e);
  }

  @Override
  public boolean onSingleTapConfirmed(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onSingleTapConfirmed(e);
    }
    return mDefaultGestureHandler.onSingleTapConfirmed(e);
  }

  @Override
  public boolean onSingleTapUp(MotionEvent e) {
    if (mOuterListener != null) {
      mOuterListener.onSingleTapUp(e);
    }
    return mDefaultGestureHandler.onSingleTapUp(e);
  }
}

File: frescophotoview/src/main/java/com/zml/frescophotoview/GestureListenerWrapper.kt
package com.zml.frescophotoview

import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent

class GestureListenerWrapper(private val defaultGestureHandler: DefaultGestureHandler) :
    SimpleOnGestureListener() {
    private var mOuterListener: SimpleOnGestureListener? = null
    fun setListener(listener: SimpleOnGestureListener?) {
        mOuterListener = listener
    }

    override fun onLongPress(e: MotionEvent) {
        mOuterListener?.onLongPress(e)
        defaultGestureHandler.onLongPress(e)
    }

    override fun onScroll(
        e1: MotionEvent,
        e2: MotionEvent,
        distanceX: Float,
        distanceY: Float
    ): Boolean {
        mOuterListener?.onScroll(e1, e2, distanceX, distanceY)
        return defaultGestureHandler.onScroll(e1, e2, distanceX, distanceY)
    }

    override fun onFling(
        e1: MotionEvent,
        e2: MotionEvent,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        mOuterListener?.onFling(e1, e2, velocityX, velocityY)
        return defaultGestureHandler.onFling(e1, e2, velocityX, velocityY)
    }

    override fun onShowPress(e: MotionEvent) {
        mOuterListener?.onShowPress(e)
        defaultGestureHandler.onShowPress(e)
    }

    override fun onDown(e: MotionEvent): Boolean {
        mOuterListener?.onDown(e)
        return defaultGestureHandler.onDown(e)
    }

    override fun onDoubleTap(e: MotionEvent): Boolean {
        mOuterListener?.onDoubleTap(e)
        return defaultGestureHandler.onDoubleTap(e)
    }

    override fun onDoubleTapEvent(e: MotionEvent): Boolean {
        mOuterListener?.onDoubleTapEvent(e)
        return defaultGestureHandler.onDoubleTapEvent(e)
    }

    override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
        mOuterListener?.onSingleTapConfirmed(e)
        return defaultGestureHandler.onSingleTapConfirmed(e)
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean {
        mOuterListener?.onSingleTapUp(e)
        return defaultGestureHandler.onSingleTapUp(e)
    }
}